lib/charms/mysql/v0/mysql.py

```diff
@@ -311,6 +333,253 @@ class RouterUser:
+                content = secret.get_content()
+                self.unit_secrets = content
+
+            logger.debug(f"Retrieved secret {key} for unit")
+            return self.unit_secrets.get(key)
+
+        secret_id = self.app_peer_data.get("secret-id")
+
+        if not self.app_secrets and not secret_id:
+            logger.debug("Getting a secret when no secrets added in juju")
+            return None
+
+        if not self.app_secrets:
+            secret = self.model.get_secret(id=secret_id)
+            content = secret.get_content()
+            self.app_secrets = content
+
+        logger.debug(f"Retrieved secret {key} for app")
+        return self.app_secrets.get(key)
+
+    def _get_secret_from_databag(self, scope: str, key: str) -> Optional[str]:
+        """Retrieve and return the secret from the peer relation databag."""
+        if scope == "unit":
+            return self.unit_peer_data.get(key)
+
+        return self.app_peer_data.get(key)
+
+    def get_secret(self, scope: str, key: str) -> Optional[str]:
+        """Get secret from the secret storage."""
+        if scope not in ["unit", "app"]:
+            raise MySQLSecretError(f"Invalid secret scope: {scope}")
+
+        if ops.jujuversion.JujuVersion.from_environ().has_secrets:
+            return self._get_secret_from_juju(scope, key)
+
+        return self._get_secret_from_databag(scope, key)
+
+    def _set_secret_in_databag(self, scope: str, key: str, value: str) -> None:
+        """Set secret in the peer relation databag."""
+        if not value:
+            if scope == "unit":
+                del self.unit_peer_data[key]
+            else:
+                del self.app_peer_data[key]
+            return
+
+        if scope == "unit":
+            self.unit_peer_data.update({key: value})
+            return
+
+        self.app_peer_data.update({key: value})
+
+    def _set_secret_in_juju(self, scope: str, key: str, value: str) -> None:
+        """Set the secret in the juju secret storage."""
+        if scope == "unit":
+            secret_id = self.unit_peer_data.get("secret-id")
+        else:
+            secret_id = self.app_peer_data.get("secret-id")
+
+        if secret_id:
+            secret = self.model.get_secret(id=secret_id)
+            content = secret.get_content()
+
+            if not value:
+                del content[key]
+            else:
+                content[key] = value
+
+            secret.set_content(content)
+            logger.debug(f"Updated {scope} secret {secret_id} for {key}")
+        else:
+            content = {
+                key: value,
+            }
+
+            if scope == "unit":
+                secret = self.unit.add_secret(content)
+                self.unit_peer_data["secret-id"] = secret.id
+            else:
+                secret = self.app.add_secret(content)
+                self.app_peer_data["secret-id"] = secret.id
+            logger.debug(f"Added {scope} secret {secret_id} for {key}")
```

Comment:
please add caching

Reply:
added in [27bf3db](https://github.com/canonical/mysql-operator/pull/243/commits/27bf3db7d627761cbc72a770c1adc3f80028b134)